File: 2024/Puzzles/11/Day11.cs
namespace AdventOfCode2024.Puzzles;

static class Day11
{
    const string Filename = "Puzzles/11/input.txt";
    static readonly Dictionary<(long, int), long> Cache = [];

    public static void Run()
    {
        List<long> stones = GetStones();

        long part1 = stones.CountStones(25);
        long part2 = stones.CountStones(75);

        Console.WriteLine($"** Day 11 **");
        Console.WriteLine($"Part 1: {part1}");
        Console.WriteLine($"Part 2: {part2}");
    }

    static List<long> GetStones()
    {
        List<long> stones = [];

        foreach (string line in File.ReadLines(Filename))
        {
            stones.AddRange([.. line.Split(' ').Select(long.Parse)]);
        }

        return stones;
    }

    static long CountStones(this IEnumerable<long> stones, int blinks) =>
        stones.Sum(stone => stone.CountStones(blinks));

    static long CountStones(this long stone, int blinks) =>
        blinks is 0 ? 1
        : Cache.TryGetValue((stone, blinks), out var count) ? count
        : Cache[(stone, blinks)] = stone.Blink().CountStones(blinks - 1);

    static long[] Blink(this long stone) =>
        stone is 0 ? [1]
        : stone.DigitsCount() is int count && count % 2 is 0 ? stone.Split((count / 2).Pow10())
        : [stone * 2024];

    static int Pow10(this int n) =>
        (int)Math.Pow(10, n);

    static long[] Split(this long n, int div) =>
        [n / div, n % div];
    
    static int DigitsCount(this long n) =>
        (int)Math.Log10(n) + 1;
}

File: 2016/puzzles/day11.py
# INPUT_FILE = "inputs/11.txt"


def run() -> None:
    floors = [8, 2, 0, 0]
    real_floors = [floors[0] + 4] + floors[1:]

    part1 = get_moves(floors)
    part2 = get_moves(real_floors)

    print("** Day 11 **")
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")

# TODO: do it by myself
def get_moves(items: list[int]) -> int:
    moves = 0

    while items[-1] != sum(items):
        curr_floor = 0

        while items[curr_floor] == 0:
            curr_floor += 1

        moves += 2 * (items[curr_floor] - 1) - 1
        items[curr_floor + 1] += items[curr_floor]
        items[curr_floor] = 0

    return moves
